Kafka external sharding strategy (#418)
lightbend/cloudflow: docs/shared-content-source/docs/modules/develop/pages/clustering-akka-streamlet.adoc

build stateful Akka Streamlets.

== Akka Cluster

WARNING: The features on this page feature are experimental and may change.  It is not recommended to use these features in production just yet.

=== Use Case

Akka Clustering in a Cloudflow Akka Streamlet gives you the ability to leverage existing Akka tools for managing
data in-memory. Akka Cluster Sharding can be used where data must be consistent and Akka Distributed Data
can be used for eventually consistent highly-available data.  There exist many real world use cases where
stateful streams are needed, including model updates in real-time ML serving and sharding IOT Device State.

Add the Clustering trait to your streamlet as shown below to activate Clustering on a specific Akka Streamlet.

[source,scala]
----
object ConnectedCarCluster extends AkkaStreamlet with Clustering
----

By including the Clustering trait on your Akka Streamlet, Cloudflow will automatically setup all the
configuration your streamlet needs to form a cluster, both when running the application locally and when the application is deployed to a Kubernetes cluster.

== Kafka External Sharding Source

> When using Akka Cluster Sharding with a normal cloudflow source it is likely that the location of a specific

Reviewer: ```suggestion
When using Akka Cluster Sharding with a normal Cloudflow source it is likely that the location of a specific
```

Akka shard is not the same as the node assigned to that shard's corresponding Kafka partition.  In this case every
message read from kafka would need to be sent over the network to the Shard node and back.

> To avoid this case and coordinate the location of Akka shards and Kafka partitions cloudflow users can

Reviewer: ```suggestion
To avoid this case and coordinate the location of Akka shards and Kafka partitions you can
```

> you can use `shardedSourceWithCommittable` and `shardedPlainSource`.  These Sources use an an Akka feature

Reviewer: ```suggestion
use `shardedSourceWithCommittable` and `shardedPlainSource`.  These Sources use an an Akka feature
```

called a
https://doc.akka.io/docs/alpakka-kafka/current/cluster-sharding.html[Kafka ExternalShardAllocationStrategy].

> To use these Sources you must first implement Clustering on your Akka Streamlet then follow

Reviewer: ```suggestion
To use these Sources you must first implement Clustering on your Akka Streamlet and then follow
```

the examples below.

> `shardedSourceWithCommittableContext` like its non-sharded counterpart will include Kafka offset information

Reviewer: ```suggestion
Like its non-sharded counterpart, `shardedSourceWithCommittableContext` will include Kafka offset information
```

> as context as well as implementing the Kafka Sharding strategy

Reviewer: ```suggestion
as context as well as implementing the Kafka Sharding strategy.
```

[source,scala]
----
val entity = Entity(typeKey)(createBehavior = entityContext => ConnectedCarActor(entityContext.entityId))

val source:SourceWithContext[ConnectedCarERecord, CommittableOffset, _] = shardedSourceWithCommittableContext(in, entity)
----

`shardedPlainSource` does not include any Kafka offset information as context but implements the Kafka

> Sharding Strategy in the same way

Reviewer: ```suggestion
Sharding Strategy in the same way.
```

:page-partial:
:page-supergroup-scala-java: Language

include::ROOT:partial$include.adoc[]

https://doc.akka.io/docs/akka/current/typed/cluster.html[Akka Cluster] is a feature of Akka that

> allows Akka nodes to form a cluster and communicate.  This feature can be used within cloudflow to

Reviewer: ```suggestion
allows Akka nodes to form a cluster and communicate.  This feature can be used within Cloudflow to
```

build stateful Akka Streamlets.

== Akka Cluster

WARNING: The features on this page feature are experimental and may change.  It is not recommended to use these features in production just yet.

=== Use Case

> Akka Clustering in a cloudflow Akka Streamlet gives you the ability to leverage existing Akka tools for managing

Reviewer: ```suggestion
Akka Clustering in a Cloudflow Akka Streamlet gives you the ability to leverage existing Akka tools for managing
```

> data in-memory. Akka Cluster Sharding can be used to where data must be consistent and Akka Distributed Data

Reviewer: ```suggestion
data in-memory. Akka Cluster Sharding can be used where data must be consistent and Akka Distributed Data
```

can be used for eventually consistent highly-available data.  There exist many real world use cases where

> stateful stream are needed including model updates in real-time ML serving and sharding IOT Device State.

Reviewer: ```suggestion
stateful streams are needed, including model updates in real-time ML serving and sharding IOT Device State.
```

> To activate Clustering on a specific Akka Streamlet add the Clustering trait to your streamlet as show below.

Reviewer: ```suggestion
Add the Clustering trait to your streamlet as shown below to activate Clustering on a specific Akka Streamlet.
```

[source,scala]
----
object ConnectedCarCluster extends AkkaStreamlet with Clustering
----

> By including the clustering Trait on you Akka Streamlet cloudflow will automatically setup all the

Reviewer: ```suggestion
By including the Clustering trait on your Akka Streamlet, Cloudflow will automatically setup all the
```

> configs your streamlet needs to form a cluster locally and deployed on kubernetes.

Reviewer: ```suggestion
configuration your streamlet needs to form a cluster, both when running the application locally and when the application is deployed to a Kubernetes cluster.
```
